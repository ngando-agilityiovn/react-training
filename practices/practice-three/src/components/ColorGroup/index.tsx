import { Box, BoxProps, chakra, useRadio } from '@chakra-ui/react';

interface CustomRadioProps extends BoxProps {
  color: string;
}

const ColorGroup = (props: CustomRadioProps) => {
  const { color, ...radioProps } = props;
  const { state, getInputProps, getRadioProps, htmlProps, getLabelProps } = useRadio(radioProps);

  return (
    <chakra.label {...htmlProps} cursor="pointer">
      <input {...getInputProps({})} hidden />
      <Box w="16" h="16" {...getRadioProps()} color={color} m="34px 12px 0 40px">
        <Box
          _hover={{ background: color }}
          aria-label=""
          bg={color}
          w={16}
          h={16}
          rounded="100%"
          {...getLabelProps()}
          content='""'
          top="50%"
          left="50%"
          transform="translate(-50%, -50%)"
          width="120%"
          height="120%"
          borderRadius="100%"
          _after={{
            position: 'absolute',
            content: '""',
            top: '50%',
            left: '50%',
            transform: 'translate(-50%, -50%)',
            width: '120%',
            height: '120%',
            bg: 'transparent',
            borderRadius: '100%',
            border: `4px solid  ${state.isChecked ? color : 'transparent'}`
          }}
          zIndex={-1}
        />
      </Box>
    </chakra.label>
  );
};

export default ColorGroup;
