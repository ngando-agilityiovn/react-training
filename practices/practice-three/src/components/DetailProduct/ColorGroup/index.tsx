import { Box, Flex, IconButton, Text } from '@chakra-ui/react'

interface IColorItem {
  color: string
}

interface IRadioGroup {
  data: IColorItem[]
}

const ColorGroup = ({ data }: IRadioGroup) => {
  return (
    <>
      <Text color="backgroundWarning" fontWeight="medium" mb="14px">
        Choose a Color
      </Text>
      <Flex gap="12px">
        {data.map(({ color }) => (
          <Box color={color}>
            <IconButton
              aria-label=""
              bg={color}
              w={30}
              h={30}
              rounded="full"
              position="relative"
              _after={{
                position: 'absolute',
                content: '""',
                top: '50%',
                left: '50%',
                transform: 'translate(-50%, -50%)',
                width: '120%',
                height: '120%',
                bg: 'transparent',
                borderRadius: '100%',
                border: `4px solid  ${color}`,
                zIndex: -1,
              }}
            />
          </Box>
        ))}
      </Flex>
    </>
  )
}

export default ColorGroup
