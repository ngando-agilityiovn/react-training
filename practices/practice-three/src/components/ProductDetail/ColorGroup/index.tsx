import { Box, Flex, IconButton, Text } from '@chakra-ui/react'
import { useState } from 'react'

interface IColorItem {
  color: string
}

interface IRadioGroup {
  data: IColorItem[]
}

const ColorGroup = ({ data }: IRadioGroup) => {
  const [isSelected, setIsSelected] = useState(0)
  return (
    <>
      <Text color="backgroundWarning" fontWeight="medium" mb="14px">
        Choose a Color
      </Text>
      <Flex gap="6">
        {data?.map(({ color }, index) => {
          return (
            <Box color={color} key={index}>
              <IconButton
                _hover={{ background: color }}
                aria-label=""
                bg={color}
                w={16}
                h={16}
                rounded="100%"
                position="relative"
                onClick={() => setIsSelected(index)}
                // TODO: Implement handle after
                _after={{
                  position: 'absolute',
                  content: '""',
                  top: '50%',
                  left: '50%',
                  transform: 'translate(-50%, -50%)',
                  width: '120%',
                  height: '120%',
                  bg: 'transparent',
                  borderRadius: '100%',
                  border: `4px solid  ${
                    isSelected === index
                      ? data[isSelected].color
                      : 'transparent'
                  }`,
                  zIndex: -1
                }}
              />
            </Box>
          )
        })}
      </Flex>
    </>
  )
}

export default ColorGroup
