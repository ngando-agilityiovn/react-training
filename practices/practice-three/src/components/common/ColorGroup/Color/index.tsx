import { Box, BoxProps, chakra, useRadio } from '@chakra-ui/react';
import { memo } from 'react';

interface CustomRadioProps extends BoxProps {
  color: string;
}

const Color = (props: CustomRadioProps) => {
  const { color, ...radioProps } = props;
  const { state, getInputProps, getRadioProps, htmlProps, getLabelProps } =
    useRadio(radioProps);

  return (
    <chakra.label aria-label={`radio-${color}`} {...htmlProps} cursor="pointer">
      <input {...getInputProps({})} />
      <Box
        w="16"
        h="16"
        {...getRadioProps()}
        color={color}
        m="34px 12px 0 40px"
      >
        <Box
          _hover={{ background: color }}
          bg={color}
          w={16}
          h={16}
          rounded="100%"
          {...getLabelProps()}
          content='""'
          top="50%"
          left="50%"
          transform="translate(-50%, -50%)"
          width="120%"
          height="120%"
          _after={{
            position: 'absolute',
            content: '""',
            top: '50%',
            left: '50%',
            transform: 'translate(-50%, -50%)',
            width: '120%',
            height: '120%',
            bg: 'transparent',
            borderRadius: '100%',
            border: `4px solid ${state.isChecked ? color : 'transparent'}`
          }}
          zIndex={-1}
        />
      </Box>
    </chakra.label>
  );
};

export default memo(Color);
