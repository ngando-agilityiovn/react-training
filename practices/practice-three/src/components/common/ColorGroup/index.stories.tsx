import { Meta, StoryObj } from '@storybook/react'
import { BrowserRouter } from 'react-router-dom'

// Constants
import { DATA_COLOR } from '@/constants'

// Components
import ColorGroup from '.'

export default {
  title: 'Components/ColorGroup',
  component: ColorGroup,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    )
  ]
} as Meta

type Story = StoryObj<typeof ColorGroup>

export const Default: Story = {
  args: { colors: DATA_COLOR }
}
