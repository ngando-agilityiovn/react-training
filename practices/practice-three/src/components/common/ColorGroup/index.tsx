import { HStack, Stack, Text, useRadioGroup } from '@chakra-ui/react'
import Color from './Color'

interface IColorGroup {
  colors?: string[]
}

const ColorGroup = ({ colors }: IColorGroup) => {
  const { getRadioProps, getRootProps } = useRadioGroup({
    defaultValue: '#3498db'
  })

  return (
    <Stack {...getRootProps()} borderBottom="1px solid gainsboro">
      <Text color="backgroundWarning" fontWeight="medium" mb="14px">
        Choose a Color
      </Text>
      <HStack>
        {colors?.map((item) => {
          return (
            <Color
              key={item}
              color={item}
              {...getRadioProps({ value: item })}
            />
          )
        })}
      </HStack>
    </Stack>
  )
}

export default ColorGroup
