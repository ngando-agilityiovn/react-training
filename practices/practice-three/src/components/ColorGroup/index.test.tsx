import { render } from '@testing-library/react';
import { useRadio } from '@chakra-ui/react';

// Components
import ColorGroup from '.';

jest.mock('@chakra-ui/react', () => ({
  ...jest.requireActual('@chakra-ui/react'),
  useRadio: jest.fn()
}));
describe('ColorGroup component', () => {
  it('renders Color component correctly', () => {
    const color = 'blue';

    const mockUseRadio = jest.fn();
    mockUseRadio.mockReturnValue({
      state: { isChecked: true },
      getInputProps: jest.fn(),
      getRadioProps: jest.fn(),
      htmlProps: {},
      getLabelProps: jest.fn()
    });
    (useRadio as jest.Mock).mockImplementation(mockUseRadio);

    const { container } = render(<ColorGroup color={color} />);
    expect(container).toMatchSnapshot();
  });
});
