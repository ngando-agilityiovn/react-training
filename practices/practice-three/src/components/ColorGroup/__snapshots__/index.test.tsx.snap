// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorGroup component renders Color component correctly 1`] = `
<div>
  <div
    class="chakra-stack css-8g8ihq"
    role="radiogroup"
  >
    <p
      class="chakra-text css-1f7e0rw"
    >
      Choose a Color
    </p>
    <div
      class="chakra-stack css-1igwmid"
    >
      <label
        class="css-4g6ai3"
      >
        <input
          hidden=""
        />
        <div
          class="css-dmcr11"
        >
          <div
            aria-label=""
            class="css-1icq76o"
            content=""""
          />
        </div>
      </label>
      <label
        class="css-4g6ai3"
      >
        <input
          hidden=""
        />
        <div
          class="css-90x6df"
        >
          <div
            aria-label=""
            class="css-1h4dol3"
            content=""""
          />
        </div>
      </label>
      <label
        class="css-4g6ai3"
      >
        <input
          hidden=""
        />
        <div
          class="css-10lllrq"
        >
          <div
            aria-label=""
            class="css-1h97wqx"
            content=""""
          />
        </div>
      </label>
      <label
        class="css-4g6ai3"
      >
        <input
          hidden=""
        />
        <div
          class="css-1bxy0tl"
        >
          <div
            aria-label=""
            class="css-v8t28h"
            content=""""
          />
        </div>
      </label>
      <label
        class="css-4g6ai3"
      >
        <input
          hidden=""
        />
        <div
          class="css-18na94k"
        >
          <div
            aria-label=""
            class="css-1ge6590"
            content=""""
          />
        </div>
      </label>
    </div>
  </div>
</div>
`;
